src/mgmt/srv_pool.c

```diff
@@ -494,13 +494,34 @@ ds_mgmt_create_pool(uuid_t pool_uuid, const char *group, char *tgt_dev,
 	struct mgmt_svc			*svc;
 	d_rank_list_t			*rank_list;
 	uuid_t				*tgt_uuids = NULL;
+	d_rank_list_t			*filtered_targets = NULL;
+	d_rank_list_t			*pg_ranks = NULL;
+	uint32_t			pg_size;
 	int				rc;
 	int				rc_cleanup;
 
 	rc = ds_mgmt_svc_lookup_leader(&svc, NULL /* hint */);
 	if (rc != 0)
 		goto out;
 
+	/* Sanity check targets versus cart's current primary group members.
+	 * If any targets not in PG, flag error before MGMT_TGT_ corpcs fail.
+	 */
+	rc = crt_group_size(NULL, &pg_size);
+	D_ASSERTF(rc == 0, ""DF_RC"\n", DP_RC(rc));
+	pg_ranks = d_rank_list_alloc(pg_size);
+	rc = d_rank_list_dup(&filtered_targets, targets);
+	if (rc) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+	/* Remove any targets not found in pg_ranks */
+	d_rank_list_filter(pg_ranks, filtered_targets, false /* exclude */);
+	if (!d_rank_list_identical(filtered_targets, targets)) {
+		D_ERROR("some ranks not found in cart primary group\n");
+		D_GOTO(out, rc = -DER_OOG);
+	}
+
```

Comment:
Does this problem occur only with the new Go MS or also with the current C MS?

The C MS guarantees that the local crt primary group on the service leader always matches the system map. (If an error occurs when updating the leader's local crt primary group, the leader steps down.) Hence, it just needs to check if the incoming rank list is a subset of the system map in pool_create_prepare.

The new Go MS could use the same algorithm.

Reply:
The C MS suffers from the following issue for which a check like this has been found to be helpful. From DAOS-5531 comment:

on master today, if a client (e.g., a negative test) were to create a pool with a specified rank list and if that rank list intentionally includes a rank that is not part of the daos system, then the implementation also returns -DER_OOG to the client, and also exhibits similar MGMT_TGT_CREATE and MGMT_TGT_DESTROY (in pool create error unwind execution flow) corpc behavior.

Reply:
With the new MS, there is a small window during which a rank or ranks can be seen as joined to the system, but have not yet been added to the crt group. The result of this is that if a pool create request comes in during that window, then the request is likely to contain ranks that are seen as joined by the MS but are not yet known to crt. This is an artifact of decoupling the MS from the ioserver, as there is looser coordination between the two now.

The solution I developed for this is to just make -DER_OOG a retryable pool create error. After 250-500ms, the group is updated and the pool create succeeds. I discovered that this solution didn't work well after Ken's work for DAOS-5206 landed, because the retries induced multiple cleanup attempts that were themselves failing. Ken developed this patch to fix that problem by proactively detecting when a pool create is likely to fail due to -DER_OOG and just returning early before attempting to create targets, etc.

Reply:
What would the problem be if the new Go MS also updates the MS leader's local crt primary group synchronously via dRPC? We wouldn't need to do the rank filtering in every dRPC handler that does CoRPCs.

Reply:
I spent a couple of days on this idea. In the end, I think that it is the correct approach, but I ran into a lot of startup sequencing issues with my new implementation. For the moment, I want to push forward with the current approach and then revisit this after the main PR has landed. There are two main problems to be resolved with making the group update synchronous: 1) The initial bootstrap is more complicated because the MS has to have running ranks in order to call group update (but if it has to call group update in order to complete join for those ranks, then we're going in circles) and 2) because we lost the C-based ms rdb, the map broadcast call was pulled into the group update call in order to take advantage of the server list supplied to group update.

I think that I should probably pull the map broadcast back out, perhaps by creating a minimal svc implementation that caches the server list received on group update. That way we can continue to use the existing code for doing an async map broadcast (with its retry logic, etc).

Reply:
@mjmac, thanks for giving this a thought.

For problem 1, we could try to come up with an abstraction of the local daos_io_server. For example, if we call this abstraction foo, then we could conceive this model:

- An MS can perform certain functionalities (e.g., pool creation, system map distibution) iff foo is available;
- Initializing foo involves synchronizing its with the latest system map;
- The system map distribution shall make attempts only when foo is available and stop when foo is unavailable.

For problem 2, the Go MS need to do two different things:

- Give foo the latest system map to update the local group.
-- The local daos_io_server calls crt_group_primary_modify.
- Give foo the latest system map to update the group on all servers.
-- The local daos_io_server does map_update_bcast.

Implementing them with one dRPC/function---I imagine that's how you do it---makes sense to me. It's also quite reasonable to drive these actions from Go via dRPCs. Or, am I missing the point here? :)